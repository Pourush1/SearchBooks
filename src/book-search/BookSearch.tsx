import React, { useEffect, useState } from 'react';
import { getBooksByType } from './book-search.service';
import BookList from '../components/BookList/BookList';
import WishList from '../components/WishList/WishList';

export interface IBook {
  title: string;
  authors?: string[];
  publishedDate?: string;
  thumbnail?: string;
  publisher?: string;
  description?: string;
}

const BookSearch = () => {
  const [bookType, updateBookType] = useState('');
  const [bookTypeToSearch, updateBookTypeToSearch] = useState('');
  const [allAvailableBooks, setAllAvailableBooks] = useState<IBook[]>([]);
  const [wishList, setWishList] = useState<IBook[]>([]);
  const [isLoading, setLoading] = useState(false);
  async function requestBooks() {
    if (bookTypeToSearch) {
      setLoading(true);
      const { items } = await getBooksByType(bookTypeToSearch);
      if (items) {
        const requiredBooks: IBook[] = items.map((book: any) => {
          const { volumeInfo } = book;
          return {
            title: volumeInfo.title,
            authors: volumeInfo.authors,
            publishedDate: volumeInfo.publishedDate,
            thumbnail: volumeInfo?.imageLinks?.thumbnail,
            publisher: volumeInfo.publisher,
            description: volumeInfo.description
          };
        });
        setAllAvailableBooks(requiredBooks);
        setLoading(false);
      } else {
        setLoading(false);
      }
    }
  }

  useEffect(() => {
    async function getAllBooks() {
      await requestBooks();
    }
    getAllBooks();
  }, [bookTypeToSearch]);

  useEffect(() => {
    if (bookType) {
      const timeOut = setTimeout(() => updateBookTypeToSearch(bookType), 500);
      return () => clearTimeout(timeOut);
    } else {
      setAllAvailableBooks([]);
    }
  }, [bookType]);

  const addToWishList = (book: IBook) => {
    let newWishList = [...wishList];
    let itemInCart = newWishList.find(
      item => item.thumbnail === book.thumbnail
    );
    if (!itemInCart) {
      let newItem = {
        title: book.title,
        publisher: book.publisher,
        thumbnail: book.thumbnail
      };
      newWishList.push(newItem);
    }
    console.log(newWishList);
    setWishList(newWishList);
  };

  return (
    <>
      <div className="book--container">
        <div className="search-params">
          <div>
            <form
              onSubmit={e => {
                e.preventDefault();
                updateBookTypeToSearch(bookType);
              }}
            >
              <input
                className="full-width"
                autoFocus
                name="gsearch"
                type="search"
                value={bookType}
                placeholder="Search for books to add to your reading list and press Enter"
                onChange={e => updateBookType(e.target.value)}
              />
            </form>

            {!bookType ? (
              <div className="empty">
                <p>
                  Try searching for a topic, for example
                  <a
                    onClick={() => {
                      updateBookType('Javascript');
                    }}
                  >
                    {' '}
                    "Javascript"
                  </a>
                </p>
              </div>
            ) : (
              isLoading && (
                <div className="d-flex text-center justify-content-center">
                  <div className="spinner-border d-flex" role="status"></div>
                </div>
              )
            )}
          </div>
        </div>
      </div>

      {allAvailableBooks.length > 0 && (
        <div className="row flex-md-row-reverse mx-3 ">
          <div className="col-md-4  border border-primary m-3 m-md-0">
            <WishList books={wishList} />
          </div>
          <div className="col-md-8">
            <BookList books={allAvailableBooks} addToWishList={addToWishList} />
          </div>
        </div>
      )}
    </>
  );
};

export default BookSearch;
