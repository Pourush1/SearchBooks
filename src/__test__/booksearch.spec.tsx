import React from 'react';
import {
  render,
  RenderResult,
  screen,
  fireEvent
} from '@testing-library/react';
import BookSearch from '../book-search/BookSearch';

let documentBody: RenderResult;

describe('<BookSearch />', () => {
  beforeEach(() => {
    documentBody = render(<BookSearch />);
  });

  it('renders initial BookSearch Component', () => {
    expect(
      screen.getByText(`Try searching for a topic, for example`)
    ).toBeInTheDocument();
    expect(screen.getByText(/Try searching for a topic/)).toBeInTheDocument();
    expect(screen.getByRole('searchbox')).toBeInTheDocument();
    expect(
      screen.getByPlaceholderText(
        'Search for books to add to your reading list and press Enter'
      )
    ).toBeInTheDocument();
    expect(screen.queryByText(/Javascript/)).toBeInTheDocument();
    // screen.debug();
    // screen.getByRole('');
  });

  it('assert for elements that is not there', () => {
    expect(screen.queryByText(/Find for Javascript/)).toBeNull();
  });

  it('matches snapshot', () => {
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });
});
