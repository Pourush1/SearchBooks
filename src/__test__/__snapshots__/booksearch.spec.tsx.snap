// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookSearch /> matches snapshot 1`] = `
<body>
  <div>
    <div
      class="book--container"
    >
      <div
        class="search-params"
      >
        <div>
          <form>
            <input
              class="full-width"
              name="gsearch"
              placeholder="Search for books to add to your reading list and press Enter"
              type="search"
              value=""
            />
          </form>
          <div
            class="empty"
          >
            <p>
              Try searching for a topic, for example
              <a>
                 
                "Javascript"
              </a>
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
